[plugin-excel] Add ability to specify cells type during data creation (#5146)
vividus-framework/vividus: docs/modules/plugins/partials/excel-cell-types.adoc

== Excel cell types

[cols="1,3,3", options="header"]
|===

|Type
|Usage example
|Description

|STRING
|`String value`

`#{withCellType(String, String value)}`
|Default cells type. Used if the cell type is not specified explicitly

|FORMULA
|`\#{withCellType(Formula, DOLLAR(A2))}`

`#{withCellType(Formula, A2+A3)}`
|https://support.microsoft.com/en-us/office/data-types-in-data-models-e2388f62-6122-4e2b-bcad-053e3da9ba90[Excel formulas]

|DATE
|`\#{withCellType(Date, 03/31/2024)}` - date with default date pattern `MM/dd/yyyy` and default cell format `m/d/yy`. Output `3/31/24`

> `#{withCellType(Date, 31-Mar-2024 pattern dd-MMM-yyyy)}` - date with custom date pattern `dd-MMM-yyyy` and default cell format `m/d/yy`. Output `3/31/24`

Reviewer: ```suggestion
`#{withCellType(Date, 31-Mar-2024 input dd-MMM-yyyy)}` - date with custom date input format `dd-MMM-yyyy` and default cell format `m/d/yy`. Output `3/31/24`
```

> `\#{withCellType(Date, 03/31/2024 format m.d.yyyy h:mm)}` - date with default date pattern `MM/dd/yyyy` and custom cell format `m.d.yyyy h:mm`. Output `3.31.2024 0:00`

Reviewer: ```suggestion
`\#{withCellType(Date, 03/31/2024 format m.d.yyyy h:mm)}` - date with default date input format `MM/dd/yyyy` and custom cell format `m.d.yyyy h:mm`. Output `3.31.2024 0:00`
```

> `#{withCellType(Date, 2024/03/31 13:04 pattern yyyy/MM/dd HH:mm format m.d.yy h:mm)}` - date with custom date and cell formats. Output `3.31.24 13:04`

Reviewer: ```suggestion
`#{withCellType(Date, 2024/03/31 13:04 input yyyy/MM/dd HH:mm format m.d.yy h:mm)}` - date with custom date input and cell formats. Output `3.31.24 13:04`
```

|Excel date.

Optional parameters:

> `pattern` - the current date pattern based on the https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/time/format/DateTimeFormatter.html#patterns[date/time format]. Default value `MM/dd/yyyy`

Reviewer: ```suggestion
`input` - the input date format based on the https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/time/format/DateTimeFormatter.html#patterns[date/time format]. Default value is `MM/dd/yyyy`
```

> `format` - the current https://support.microsoft.com/en-us/office/format-numbers-as-dates-or-times-418bd3fe-0577-47c8-8caa-b4d30c528309[cell date format] in Excel. Default value `m/d/yy`

Reviewer: ```suggestion
`format` - the output https://support.microsoft.com/en-us/office/format-numbers-as-dates-or-times-418bd3fe-0577-47c8-8caa-b4d30c528309[cell date format] in Excel. Default value `m/d/yy`
```

|BOOLEAN
|`#{withCellType(Boolean, false)}`
|Boolean cell type

|NUMERIC
|`\#{withCellType(Numeric, 128)}` - output `128.0`

`#{withCellType(Numeric, 32768 format \#,##0.00)}` - the number with `,` thousands separator and two decimal places. Output `32,768.00`
|Number cell type

Optional parameters:

`format` - the current https://support.microsoft.com/en-us/office/number-format-codes-5026bbd6-04bc-48cd-bf33-80f18b4eae68[cell number format]. Default value `#.0`

|===

=== Cells type for column

Cells type can be specified for all values in column using the next column name format: `name "<column-name>", type "<values-type>"`. The types specified in individual cells will have priority.

.Create Excel file with typed column
----
When I create temporary excel file with content:

> |name "multi-type column", type "date"                        |

Reviewer: ```suggestion
|#{withColumnCellsType(date, multi-type column)                        |
```

> |31-Mar-2024 pattern dd-MMM-yyyy                              |

Reviewer: ```suggestion
|31-Mar-2024 input dd-MMM-yyyy                              |
```
